### ISSUE TITLE
Crash when interactively adding a number of subplots

### ISSUE THREAD

**Issue by mdovgialo** [matplotlib/matplotlib#11123]
### Bug report

**Bug summary**

Adding subplots repeatedly to a figure in interactive mode can sometimes crash Matplotlib version 2.2.2, in version 2.0.2 everything works. Sometimes it does not crash, but when you resize figure's window to make it significantly smaller it crashes every time. I'm using Matplotlib on Kubuntu 16.04 64 bit with Python 3.5.2.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
import time

plt.ion()

def test_subplot_crash_with_matplotlib222():
    plt.ion()

    import matplotlib
    print("Matplotlib ver", matplotlib.__version__)
    fig = plt.figure()
    fig.set_size_inches(10, 10)
    for i in range(100):
        ax = fig.add_subplot(10, 10, i+1)
        ax.plot([0, 1], [0, 1])
        fig.tight_layout()
        fig.canvas.flush_events()
        print(i)
        time.sleep(0.01)
        if i == 25:
            fig.set_size_inches(3, 3)
    plt.ioff()
    plt.show()

if __name__ == '__main__':
    test_subplot_crash_with_matplotlib222()
```

**Actual outcome**

While adding 25th subplot it crashes. You can crash it earlier if you manually resize the window to make it smaller.
```
Traceback (most recent call last):
  File "tests/test_crash_matplotlib222.py", line 28, in <module>
    test_subplot_crash_with_matplotlib222()
  File "tests/test_crash_matplotlib222.py", line 18, in test_subplot_crash_with_matplotlib222
    ax = fig.add_subplot(10, 10, i+1)
  File "/home/marian/.local/lib/python3.5/site-packages/matplotlib/figure.py", line 1239, in add_subplot
    a = subplot_class_factory(projection_class)(self, *args, **kwargs)
  File "/home/marian/.local/lib/python3.5/site-packages/matplotlib/axes/_subplots.py", line 77, in __init__
    self._axes_class.__init__(self, fig, self.figbox, **kwargs)
  File "/home/marian/.local/lib/python3.5/site-packages/matplotlib/axes/_base.py", line 488, in __init__
    raise ValueError('Width and height specified must be non-negative')
ValueError: Width and height specified must be non-negative

```

**Expected outcome**
Interactive matplotlib window with dynamically added 100 plots. While plots are being added window should be resizable. After 25'th plot it will make itself smaller in an effort to reproduce the bug. In Matplotlib version 2.0.2 it works without crashing.

**Matplotlib version**

  * Operating system: Kubuntu 16.04 64 bit, Linux  4.13.0-38-generic #43~16.04.1-Ubuntu SMP Wed Mar 14 17:48:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
  * Matplotlib version: 2.2.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg
  * Python version: 3.5.2

Installed matplotlib using:
`pip3 install --user matplotlib==2.2.2`

**Comment by jklymak** [comment]
What happens if you turn off tight layout?  

What does this look like in non crashing 2.0.2?

**Comment by mdovgialo** [comment]
Hm, without tight_layout it doesn't crash in 2.2.2, but it shouldn't crash with tight_layout.

How does it look? A white window, in which plots are appearing, while the plots are appearing the window is resizable and those plots are being resized with the figure.

I'm using similar code for monitoring of online EEG signal. I add subplots on-line when new statistic is  calculated and available. My code was crashing and I've narrowed down to this minimal example. I use tight_layout for so the plot titles and labels wouldn't overlap and screen space usage would be optimal.

**Comment by QuLogic** [comment]
Pre-allocating is a bit safer, it seems:
```python3
    fig, axs = plt.subplots(10, 10)
    for ax in axs.flat:
        ax.set_visible(False)
    fig.set_size_inches(10, 10)
    for i, ax in enumerate(axs.flat):
        ax.set_visible(True)
        ax.plot([0, 1], [0, 1])
        fig.tight_layout()
        fig.canvas.flush_events()
        print(i)
        time.sleep(0.01)
        if i == 25:
            fig.set_size_inches(3, 3)
```
does not crash as far as I can tell.

**Comment by mdovgialo** [comment]
Pre-allocating is safer. But when you have 3 dynamic plots using all figure space, and then you collect enough data and want to show new plot, you have to use `Axes.change_geometry ` geometry method to make existing plots smaller and then add new subplot with newly calculated data. For now I'm staying with matplotlib 2.0.2

**Comment by jklymak** [comment]
This is fixed in master #10915.  Trying to backport to 2.2.3...  The behaviour in master is to warn when tight_layout can't make the axes small enough anymore rather than giving a ValueError

**Comment by jklymak** [comment]
Closed by #10915.  Not sure it’s exacty what you wanted but at least it doesn’t crash.  Thanks for the report.
